### PR TITLE
feat(EMS-613-637): Company details - populate fields

### PR DESCRIPTION
**Pull request by Zainzzkk**
# Changes made: 
* Populated fields on company details page in get controller if they exist
* Added helper to go map through application and populate companies house summary list
* Added tests and added to e2e test

### CODE REVIEW

File: e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/company-details-save-and-back.spec.js [hunk 1]
```diff
@@ -54,6 +54,13 @@ describe('Insurance - Your business - Company details page - Save and go back',
 
       cy.url().should('eq', `${Cypress.config('baseUrl')}${ROUTES.INSURANCE.ROOT}/${referenceNumber}${ALL_SECTIONS}`);
     });
+
+    it('should display the valid inputs on fields when going back to the page', () => {
+      cy.navigateToUrl(url);
+
+      yesRadioInput().first().should('be.checked');
+      yesRadioInput().eq(1).should('be.checked');
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/company-details-save-and-back.spec.js:62]: Could we say "shouldrender submitted values" intead of the "valid inputs"?

Also do you think it's worth having cypress selectors for each radio e.g `hasTradingNameRadio.yes` `hasTradingAddressRadio.yes`etc instead of `first()` / `eq(1)`?

File: src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * maps database response and returns body in correct format for summary list
+ * @param {object} exporterCompany
+ * @returns {CompanyHouseResponse} reformatted exporterCompany
+ */
+const mapDatabaseResponse = (exporterCompany: object) => {
+  return {
+    [COMPANY_NAME]: exporterCompany[COMPANY_NAME],
+    [COMPANY_ADDRESS]: {
+      ...exporterCompany[COMPANY_ADDRESS],
+    },
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.ts:22]: Do we need the spread operator here? Could it just be 

```js
[COMPANY_ADDRESS]: exporterCompany[COMPANY_ADDRESS],
```

File: src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+const populateCompaniesHouseSummaryList = (exporterCompany: object) => {
+  if (!exporterCompany[COMPANY_NUMBER]) {
+    return null;
+  }
+
+  const companyDetails = mapDatabaseResponse(exporterCompany);
+
+  return companyHouseSummaryList(companyDetails);
+};
+
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.ts:43]: Could you check out this `helpers/summary-lists/answers-summary-list`, i'm working on the summary lists and trying to align everything.

This answers summary list is the same requirement we have for the other summary lists we'll be building soon; Answers summary list has very similar to this - a function that maps the data to content and is in `helpers/data-content-mappings`.

Currently in the "check answers" GET controller, summary list is generated like this:

```js
const answers = mapAnswersToContent(req.session.submittedData.quoteEligibility);

const summaryList = answersSummaryList(answers);
```


What do you think? Should we continue with the way answers are mapped and passed into summary list? Or we could do this as part of the `answersSummaryList` function

Either way I think the answers mapping to content should be in `helpers/data-content-mappings` and should name it similar to `mapAnswersToContent`; There is another one called `mapQuoteToContent`

File: src/ui/server/controllers/insurance/business/company-details/index.test.ts [hunk 4]
```diff
@@ -56,15 +60,59 @@ describe('controllers/insurance/business/companies-details', () => {
   });
 
   describe('get', () => {
-    it('should render the company-details template with correct variables', () => {
+    it('should render the company-details template with correct variables if application has populated exporterCompany data', () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/index.test.ts:63]: Can we wrap the conditional in a `describe`? Easier to read IMO e.g

```js
describe('When the application has populated exporterCompany data', () => {
  it('should....
});
```
Same below

File: src/ui/server/controllers/insurance/business/company-details/index.ts [hunk 5]
```diff
@@ -48,22 +49,41 @@ const exitReason = {
  * gets the template for company details page
  * @param {Express.Request} Express request
  * @param {Express.Response} Express response
- * @returns {Express.Response.render} renders company details page
+ * @returns {Express.Response.render} renders company details page with/without previously submitted details
  */
 const get = async (req: Request, res: Response) => {
-  const { application } = res.locals;
+  try {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/index.ts:55]: Seems that we can remove `async` 😊

File: src/ui/server/controllers/insurance/business/map-submitted-data/mocks/index.ts [hunk 6]
```diff
@@ -11,7 +11,7 @@ export const mockBody = {
   [INPUT]: '8989898',
   __typename: 'CompaniesHouseResponse',
   ...mockCompany,
-  registeredOfficeAddress: {
+  address: {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/map-submitted-data/mocks/index.ts:14]: Shouldn't this be `registeredOfficeAddress` ?

File: src/ui/server/helpers/mappings/map-sic-codes.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+          sicCode: '12345',
+          __typename: 'ExporterCompanySicCode',
+        } as SicCodeResponse,
+      ];
+
+      const response = mapSicCodes(sicCodes);
+
+      expect(response).toEqual([sicCodes[0].sicCode, sicCodes[1].sicCode]);
+    });
+  });
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-sic-codes.test.ts:24]: Nice, could we move the mock sic codes into `test-mocks`?

Would be nice to have the `toEqual([...` definition into an `expected` constant so it's easier to read

File: src/ui/server/controllers/insurance/business/map-submitted-data/mocks/index.ts (hunk 6 above)
Comment (Zainzzkk) [reply src/ui/server/controllers/insurance/business/map-submitted-data/mocks/index.ts:14]: no, this is a temporary address which is processed and is passed to the api as `address` as `registeredOfficeAddress` is deleted as it contains nulls and unwanted fields like `__typename`

File: src/ui/server/helpers/summary-lists/company-house-summary-list.ts [hunk 8]
```diff
@@ -22,7 +22,7 @@ const generateAddressHTML = (address: object) => {
 
   Object.keys(address).forEach((field) => {
     // if the address field exists and not the typename part
-    if (address[field] && field !== '__typename') {
+    if (address[field] && field !== '__typename' && field !== 'id') {
```
Comment (ttbarnes) [inline src/ui/server/helpers/summary-lists/company-house-summary-list.ts:25]: Do we need to update the unit test?

File: src/ui/types/company-house-response.d.ts [hunk 9]
```diff
@@ -15,8 +15,8 @@ interface CompanyHouseResponse {
   companyNumber: string;
   dateOfCreation: string;
   sicCodes: Array<string>;
-  success: boolean;
-  apiError: boolean;
+  success?: boolean;
+  apiError?: boolean;
```
Comment (ttbarnes) [inline src/ui/types/company-house-response.d.ts:19]: What's the reason for this change?

Comment (Zainzzkk) [reply src/ui/types/company-house-response.d.ts:19]: the mapping from the database comes as  CompanyHouseResponse which does not have success or apiError and did not want to set a value from them.  Also used by the API which has those flags

Comment (ttbarnes) [reply src/ui/types/company-house-response.d.ts:19]: Could we use `ApplicationExporterCompany` type instead for these instances?

File: e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/company-details-save-and-back.spec.js (hunk 1 above)
Comment (Zainzzkk) [reply e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/company-details-save-and-back.spec.js:62]: yes, but separate ticket please as this will need a few changes

File: src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.ts (hunk 3 above)
Comment (Zainzzkk) [reply src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.ts:43]: I think the check answers approach is correct, but I think this should have a separate ticket to get working

File: src/ui/types/company-house-response.d.ts (hunk 9 above)
Comment (Zainzzkk) [reply src/ui/types/company-house-response.d.ts:19]: I would need to remap the api response, do I do it in this ticket?

Comment (ttbarnes) [reply src/ui/types/company-house-response.d.ts:19]: I think so if it's not a big job, to avoid introducing tech debt

File: e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/company-details-save-and-back.spec.js (hunk 1 above)
Comment (ttbarnes) [reply e2e-tests/cypress/e2e/journeys/insurance/your-business/company-details/company-details-save-and-back.spec.js:62]: https://ukef-dtfs.atlassian.net/browse/EMS-843

File: src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.test.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    it('should map exporterCompany to required format', () => {
+      const exporterCompany = {
+        ...mockApplication.exporterCompany,
+      };
+
+      const response = mapDatabaseResponse(exporterCompany);
+
+      const expected = {
+        ...exporterCompany,
+        [COMPANY_SIC]: [exporterCompany[COMPANY_SIC][0].sicCode],
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/business/company-details/helpers/populate-companies-house-summary-list.test.ts:53]: This should be asserted with `mapSicCodes` as that's what the code is doing, if we use `mapSicCodes` here and that function changes, won't won't need to update this test 

Not blocker for this PR

File: src/ui/types/company-details.d.ts [hunk 11]
```diff
@@ -12,4 +12,9 @@ interface CompanyDetailsFieldGroups {
   COMPANY_DETAILS: Array<SummaryListItemData>;
 }
 
-export { CompanyDetails, CompanyDetailsFieldGroups };
+interface SicCodeResponse {
+  id: string;
+  sicCode: string;
+}
```
Comment (ttbarnes) [inline src/ui/types/company-details.d.ts:18]: Can we just call this `SicCode`?

**Review by ttbarnes — DISMISSED** [review]

**Review by ttbarnes — APPROVED** [review]
